cs130-w21/1
# Implement zero-config network discovery and mock client/daemon pair

This uses the Zeroconf protocol to find peers on the LAN. The version of Zeroconf being used here is Bonjour. I have tested it locally using Windows.

To put this in a practical context, this PR also adds a dummy client-server pair. These use HTTP/2 for simplicity. The server is very dumb and will probably be totally replaced. On other hand, the client comes with a decent implementation of sending jobs from a queue to multiple servers while more servers dynamically join, so it can be the basis of a more complete implementation.

## How to test
In any order:
1. Run as many servers as you want (`npm start`), either on the same machine or different machines on the same LAN.
2. Run a client (`npm run cli`). It should find all the servers you started (log their service info to the console). Then, it should finish all 5 of its tasks, while server consoles will show that different servers got the requests.

![Sample output](https://cdn.discordapp.com/attachments/796812456500527120/802729663256461342/Screenshot_186.png)

## Interesting features
- The Zeroconf code is deliberately kept separate from the client/server.
- The daemon doesn't specify a port to listen on, so the OS assigns it a port. This makes it possible to run multiple daemons on one machine and do local testing. And it makes it so you *have* to use Zeroconf to know where the server is 😀

## Misc
This PR also adds a documentation generator (because I added doc-strings to the methods). You can view it in `docs/index.html`.

## Next steps
- Test using daemons on multiple physical machines.
- Write integration tests.
